uom/common/container/kdtree.h

```diff
@@ -0,0 +1,627 @@
+  void Rebuild(NodeUniquePtr* node);
+
+  // Flattens the sub-tree into the given array.
+  void FlattenRecursive(const NodeUniquePtr& node, std::vector<PointType>* output) const;
+
+  // Radius search from a node recursively.
+  void RadiusSearchRecursive(const NodeUniquePtr& node,
+                             const PointType& query_point,
+                             double search_radius,
+                             std::vector<PointType>* result) const;
+
+  // Helper functions for recursively print the tree.
+  void PrintTreeRecursive(const std::string& prefix,
+                          const NodeUniquePtr& node,
+                          bool is_left,
+                          bool is_deleted) const;
+
+  template <typename Callback>
+  void PreOrderTraversalRecursive(int level, TreeNode* node, Callback&& callback) const;
+
+  KdTreeParams params_;
+
+  NodeUniquePtr root_;
+
+  FRIEND_TEST(KdTreeTest, FindBestSplitAxisAndSortByMaxSpan);
+
+  DISALLOW_COPY_MOVE_AND_ASSIGN(KdTree);
+};
+
+template <typename PointType>
+struct KdTree<PointType>::Node {
+  // The point stored at *this.
+  PointType point = PointType::Zero();
+  // The division axis of this Node for kd-tree.
+  int axis = 0;
+  // The number of total nodes rooted at this node, including itself.
+  int size = 1;
+  // The number of total invalid nodes (being deleted).
+  int num_invalid_nodes = 0;
+  // The cover range of sub-tree rooted at *this.
+  RangeType range;
+  // The radius from center of sub-tree rooted at *this to the most edge.
+  double radius = 0.0;
+  // True if the sub-tree rooted at *this is deleted.
+  bool tree_deleted = false;
+  // True if *this is deleted.
+  bool point_deleted = false;
+};
+
+template <typename PointType>
+struct KdTree<PointType>::TreeNode : Node {
+  // The left child of *this.
+  NodeUniquePtr left = nullptr;
+  // The right child of *this.
+  NodeUniquePtr right = nullptr;
+  // The parent of *this, not owned.
+  TreeNode* parent = nullptr;
+
+  // Gether structure information from left and right sub-trees.
+  void PushUp();
+  // Apply structure information to left and right sub-trees.
+  void PushDown();
+  // Returns true if the node hasn't childs.
+  bool IsLeaf() const { return left == nullptr && right == nullptr; }
+  // Returns the distance from node range to given point.
+  double RangeTo(const PointType& query_point) const;
+};
+
+template <typename PointType>
+struct KdTree<PointType>::NearestNeighborSearcher {
+  NearestNeighborSearcher(const PointType& query_point, int k) : k(k), query_point(query_point) {}
+
+  void Search(const NodeUniquePtr& node);
+
+  // A heap to maintain the nearest `k` neighbors.
+  using QueueItem = std::pair<const TreeNode*, double>;
+  struct QueueItemCompare {
+    bool operator()(const QueueItem& lhs, const QueueItem& rhs) const {
+      return lhs.second < rhs.second;
+    }
+  };
+  std::priority_queue<QueueItem, std::vector<QueueItem>, QueueItemCompare> queue;
```

Comment:
IIRC, the performance  of priority_queue is lower than the implement of original heap in ikd_tree, even with the container of std::vector

Reply:
Agreed. Do you suggest me to implement a StaticPriorQueue?

Reply:
a simple array based max heap is enough

Reply:
The underlying implementation of priority queue is max heap. What's  the difference between a simple array implementation and STL？ memory pre-allocation？

I can write a prior queue based on  std::make_heap

Reply:
But `simple array` cannot support the requirement of nearest neighbors ? As the number of neighbors is unknown.

Reply:
Use a new implemented `PriorityQueue` based on `std::make_heap`. The queue is pre-allocated and reused.